File: Sort List/Sort List.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    
    private ListNode merge(ListNode l1, ListNode l2){
        ListNode fakeHead = new ListNode(0);
        ListNode curr = fakeHead;
        while(l1 != null && l2 != null){
            if(l1.val < l2.val){
                curr.next = l1;
                l1 = l1.next;
            }
            else{
                curr.next = l2;
                l2 = l2.next;
            }
            curr = curr.next;
        }
        if(l1 != null){
            curr.next = l1;
        }
        else{
            curr.next = l2;
        }
        
        return fakeHead.next;
    }
    
    public ListNode sortList(ListNode head) {
        // Length == 0 or 1
        if(head == null || head.next == null){
            return head;
        }
        
        ListNode slowPtr = head;
        ListNode fastPtr = head.next;
        while(fastPtr != null){
            fastPtr = fastPtr.next;
            if(fastPtr != null){
                fastPtr = fastPtr.next;
            }
            else{
                break;
            }
            slowPtr = slowPtr.next;
        }
        
        ListNode nextHead = slowPtr.next;
        slowPtr.next = null;
        
        head = sortList(head);
        nextHead = sortList(nextHead);
        
        ListNode newHead = merge(head, nextHead);
        
        return newHead;
    }
}

File: Sort List/Sort List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:

    ListNode* merge(ListNode* first, ListNode* second){

        if(first == NULL)           return second;
        else if(second == NULL)     return first;

        ListNode* tmpResult = NULL;

        if(first->val > second->val){
            tmpResult = second;
            tmpResult->next = merge(first, second->next);
        }
        else{
            tmpResult = first;
            tmpResult->next = merge(first->next, second);
        }
        
        return tmpResult;
    }

    ListNode* sort(ListNode* start, unsigned int len){
        if(!len)
            return NULL;
        else if(len == 1)
            return start;

        // Split the list            
        ListNode* split = start;
        unsigned int count = len / 2;
        while(count--){
            split = split->next;
        }

        // Initialize and sort first half of the list
        count = len / 2;
        ListNode* temp = start;
        while(--count){
            temp = temp->next;
        }
        temp->next = NULL;
        start = sort(start, len / 2);

        // Initialize and sort the second half of the list
        count = len - (len / 2);
        temp = split;
        while(--count){
            temp = temp->next;
        }
        temp->next = NULL;
        split = sort(split, len - (len / 2));

        ListNode* result = merge(start, split);
        return result;
    }

    ListNode *sortList(ListNode *head) {
        unsigned int len = 0;
        ListNode* tmp = head;
        for(; tmp != NULL; ++len){
            tmp = tmp->next;
        }
        return sort(head, len);
    }
};
